Rewrite and reorg of Protect documentation (#147)
ShiftLeftSecurity/documentation: using-inspect-protect/protect/json-file.md

# The ShiftLeft JSON File

> The `shiftleft.json` file is generated by ShiftLeft Inspect as part of the process of analyzing your application. This file is used by the ShiftLeft Microagent to customize the monitoring and securing of your application.  The `shiftleft.json` file is by default located in the working directory from which you run your application. 

Reviewer: by default is located in the directory you run the analysis from

Author: Done

> If you analyze your application using ShiftLeft Inspect as a [separate step before runtime](../inspect/analyzing-applications.md), the `shiftleft.json` file must be passed to ShiftLeft Protect. ShiftLeft Protect expects the file to be in the default location. If you are running your application in a different environment, copy `shiftleft.json` to the working directory of that environment.  Or you can [use the `SHIFTLEFT_CONFIG` environment variable](protect-java/configure-microagent.md) to pass to ShiftLeft Protect the path of the `shiftleft.json` file.

Reviewer 2: ```suggestion
If you analyze your application using ShiftLeft Inspect as a [separate step before runtime](../inspect/analyzing-applications.md), the `shiftleft.json` file must be passed to the ShiftLeft Microagent. The ShiftLeft Microagent expects the file to be in the default location. If you are running your application in a different environment, copy `shiftleft.json` to the working directory of that environment.  Or you can [use the `SHIFTLEFT_CONFIG` environment variable](protect-java/configure-microagent.md) to pass to the ShiftLeft Microagent the path of the `shiftleft.json` file.
```

Reviewer 2: Please update everywhere else where "Protect" is mentioned, but we should be talking about the Microagent, specifically.

Author: I think I've got all the instances

> If you analyze your application using ShiftLeft Inspect as a [separate step before runtime](../inspect/analyzing-applications.md), the `shiftleft.json` file must be passed to the Microagent. The Microagent expects the file to be in the default location. If you are running your application in a different environment, copy `shiftleft.json` to the working directory of that environment.  Or you can [use the `SHIFTLEFT_CONFIG` environment variable](protect-java/configuring-the-microagent.md) to pass to the Microagent the path of the `shiftleft.json` file.

Reviewer: looks more appropriate to put the link in "analyze your application"

Reviewer: Can we narrow down the linked URL for "SHIFTLEFT_CONFIG" to a particular HTML anchor in that page?

Author: Done

> Note that if you reanalyze your application after code changes, the `sprId` parameter changes. Since you now have a new application version, you must redeploy your application with the Microagent using the **updated** `shiftleft.json` file. 

Reviewer: Better than "redeploy" say "restart the JVM" of your application

Author: Done

> Note that if you are using a SCM system such as Git, when you reanalyze your application after code changes, the `sprId` parameter changes. Since you now have a new application version, you must redeploy your application with the Microagent using the **updated** `shiftleft.json` file. 

Reviewer 2: ```suggestion
Note that if you reanalyze your application after code changes, the `sprId` parameter changes. Since you now have a new application version, you must redeploy your application with the Microagent using the **updated** `shiftleft.json` file. 
```

Author: Done

Author: Done

> Note that if you are using a SCM system such as Git, when you reanalyze your application after code changes, the commit hash of the `sprId` parameters changes. This changed commit hash indicates that a new version of the application is built. Since you now have a new application version, you must redeploy your application with the Microagent using the **updated** `shiftleft.json` file. 

Reviewer 2: This is incorrect - we don't use the SCM / git commit hash.

The version of the built application does, however change, resulting in a new `sprId`

Author: Done

## Parameters

The `shiftleft.json` file includes the parameters `accessToken` and `sprId`, which are required by the Microagent.

### `accessToken`

The `accessToken` property represents the client access token that authorizes the Microagent to use ShiftLeft services.

### `sprId`

The `sprId` property identifies the Security Profile for Runtime (SPR) for your application. 

```json
{
  "accessToken": "${access-token-string}",
  "sprId": "${sprd-id}"
}
```

> The `sprId` is a string containing three main parts: organization ID, application name, and commit hash. For example

Reviewer 2: ```suggestion
The `sprId` is a string containing three main parts: organization ID, application name, and a unique version hash. For example
```

Note that we are using a different, internal, identifier, not the git commit hash.

Author: Done

```

> "sprId": "sl/418a892e-32fe-4d6e-b0cd-a44c24026b7a/org.springframework-my-rest-service-jar/f0e2bafa21a5790b1d70f6309189de6a1c888e16/baseline"

Reviewer 2: ```suggestion
"sprId": "sl/418a892e-32fe-4d6e-b0cd-a44c24026b7a/org.springframework-my-rest-service-jar/f0e2bafa21a5790b1d70f6309189de6a1c888e16"
```

Author: Done

Author: Done
